Patch for successive halving

stsievert opened the pull request:
Changes this PR implements:

- meta keys `ident, time_step` => `model_id, partial_fit_calls` (and have `partial_fit_calls` be the number of times `model.partial_fit` has been called)
- ~~return all the meta objects returned by `_score` instead of only some of them~~
- if `model.score` throws a `ValueError`, make sure that `meta['partial_fit_calls'] == 0`.
- renames `update` to `additional_partial_fit_calls`.

I have a successive halving implementation working correctly with this implementation.

This PR isn't formally tested, and I'm only okay submitting it because it's for https://github.com/dask/dask-ml/pull/288/.

TomAugspurger committed:
Format with black

stsievert committed:
MAINT: rename

stsievert commented:
Eh, git history is screwed up. Feel free to copy/paste the `_incremental.py` and `test_incremental.py` over – those are the only files changed.

mrocklin commented:
Out of curiosity why were these changes necessary?

Also, did the solution for successive halving in the notebook not work for
some reason?   It isn't super well explained.  Let me know if I can help
explain things there.

On Fri, Jul 27, 2018, 3:49 PM Scott Sievert <notifications@github.com>
wrote:

> Changes this PR implements:
>
>    - meta keys ident, time_step => model_id, partial_fit_calls
>    - call partial_fit the number of times specified by partial_fit_calls
>    - return all the meta objects returned by _score instead of only some
>    of them
>
> I have a successive halving implementation working correctly with this
> implementation.
>
> This PR isn't formally tested, and I'm only okay submitting it because
> it's for dask#288 <https://github.com/dask/dask-ml/pull/288>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mrocklin/dask-ml/pull/1
> Commit Summary
>
>    - use single-threaded scheduler in make_classification
>    - Add basic incremental training function
>    - Clean up incremental model selection
>    - Format with black
>    - MAINT: rename
>
> File Changes
>
>    - *M* dask_ml/datasets.py
>    <https://github.com/mrocklin/dask-ml/pull/1/files#diff-0> (4)
>    - *A* dask_ml/model_selection/_incremental.py
>    <https://github.com/mrocklin/dask-ml/pull/1/files#diff-1> (284)
>    - *A* tests/model_selection/test_incremental.py
>    <https://github.com/mrocklin/dask-ml/pull/1/files#diff-2> (134)
>
> Patch Links:
>
>    - https://github.com/mrocklin/dask-ml/pull/1.patch
>    - https://github.com/mrocklin/dask-ml/pull/1.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mrocklin/dask-ml/pull/1>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASszFETGrLucj4SwPJ2mZAjCEs6yl_oks5uK5j4gaJpZM4VkxHq>
> .
>

stsievert committed:
MAINT: make sure partial_fit_calls is proper value

stsievert committed:
MAINT: rename 'update' to 'adapt'

stsievert commented:
> Out of curiosity why were these changes necessary?

* meta keys time_step => partial_fit_calls
    * By no means required, but this makes the relationship more explicit.
*  call partial_fit the number of times specified by partial_fit_calls
    * This was very implicit; when `time_step=0`,  `model.partial_fit` had been called once. I didn't realize that until I dug into the code.
    * For my test case with successive halving, this meant that a factor of 1.2 more calls to `partial_fit` were made (486 vs 405).
* return all the meta objects returned by _score instead of only some of them
    * This is only required with a brain fart. I've removed it.

I've updated, and git history is cleaner now. It looks like you've rebased off master – thanks.

Also, I commented in https://github.com/dask/dask-ml/pull/288 about the bugs I thought I found. I deleted the comments to make diff viewing nicer.

mrocklin commented:
The `call=False` thing seems weird to me.  I'm still not sure I understand why we would call the _partial_fit function and not want to call `partial_fit`.  Are you just trying to correct for an off-by-one indexing issue?  If so then maybe its less disruptive to change an index somewhere or to change this in your code?

> For my test case with successive halving, this meant that a factor of 1.2 more calls to partial_fit were made (486 vs 405).

Can I ask you to share the code that you're building?  Also, can I ask for your thoughts on the successive-halving-ish implementation in the notebook?  Was there something lacking there that required an additional implementation?  I'm surprised that you actually needed to write an implementation here.  I was hoping that you would be able to take the existing implementations in the notebook and quickly apply them to some problem.

stsievert commented:
> Can I ask you to share the code that you're building?

https://gist.github.com/stsievert/facdf6c1427c810cafecd30848fe20d9

> can I ask for your thoughts on the successive-halving-ish implementation in the notebook? Was there something lacking there that required an additional implementation?

Nothing required additional implementation. Documentation was a little sparse, so I didn't know exactly what each value in the input meant (e.g., `time_step`). I had to dig into the code and got surprised by things that were correct but more subtle than I was expecting.

>  I'm surprised that you actually needed to write an implementation here

I only *thought* I needed a bit more functionality at the time of submitting this PR. I realized that wasn't necessary in https://github.com/mrocklin/dask-ml/pull/1#issuecomment-408574936. No new functionality is added in this PR.

I've updated https://github.com/mrocklin/dask-ml/pull/1#issue-204556255 to reflect what this PR implements.

> The call=False thing seems weird to me. I'm still not sure I understand why we would call the _partial_fit function and not want to call partial_fit.

I knew this would come up. I've removed `call` from `_partial_fit` completely, and catch `NotFittedError`s in `_score` now.

stsievert committed:
MAINT: remove call

stsievert committed:
MAINT: rename 'update' to 'get_partial_fit_calls'

stsievert committed:
DOC: document additional_partial_fit_calls

stsievert committed:
MAINT: not trained? score=0
